hazelcast/src/main/java/com/hazelcast/jet/impl/MasterJobContext.java

```diff
@@ -633,15 +669,18 @@ void cancelExecutionInvocations(long jobId, long executionId, TerminationMode mo
 
     void finalizeJob(@Nullable Throwable failure) {
         mc.coordinationService().submitToCoordinatorThread(() -> {
-            final Runnable nonSynchronizedAction;
             mc.lock();
+            JobStatus status = mc.jobStatus();
+            if (status == COMPLETED || status == FAILED) {
+                logIgnoredCompletion(failure, status);
+            }
+            mc.unlock();
+          })
+          .thenComposeAsync(r -> completeVertices(failure))
+          .thenAccept(ignored -> {
```

Comment:
The task in this `thenAccept` block must be run on the coordinator thread but currently it's not guaranteed to run on it. 
Please check: https://stackoverflow.com/a/46062939/4793261

Reply:
True, I've pushed a change that will force coordination thread executor here.